test/model/internal_link_test.dart

```diff
@@ -0,0 +1,394 @@
+      ];
+      const testCases = [
+        ('#narrow/stream/311/',      StreamNarrow(311)),
+        ('#narrow/stream/311-/',     StreamNarrow(311)),
+        ('#narrow/stream/311-help/', StreamNarrow(311)),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link', () {
+      final streams = [
+        eg.stream(streamId: 1, name: 'check'),
+        eg.stream(streamId: 2, name: 'bot testing'),
+        eg.stream(streamId: 3, name: 'check.API'),
+        eg.stream(streamId: 4, name: 'stream'),
+        eg.stream(streamId: 5, name: 'topic'),
+      ];
+      const testCases = [
+        ('#narrow/stream/check/',         StreamNarrow(1)),
+        ('#narrow/stream/bot.20testing/', StreamNarrow(2)),
+        ('#narrow/stream/check.2EAPI/',   StreamNarrow(3)),
+        ('#narrow/stream/stream/',        StreamNarrow(4)),
+        ('#narrow/stream/topic/',         StreamNarrow(5)),
+
+        ('#narrow/stream/check.API/',     null),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link, without realm info', () {
+      final stream = eg.stream(name: 'example');
+      final testCases = [
+        ('/#narrow/stream/${stream.name}/', StreamNarrow(stream.streamId)),
+        ('#narrow/stream/${stream.name}/',  StreamNarrow(stream.streamId)),
+      ];
+      checkExpectedNarrows(testCases, streams: [stream]);
+    });
+  });
+
+  group('ParseInternalLink', () {
+    group('topic link parsing', () {
+      final stream = eg.stream(name: "example");
+
+      group('basic', () {
+        String mkUrl(operand) {
+          return '#narrow/stream/${stream.streamId}-general/topic/$operand';
+        }
+        final testCases = [
+          (mkUrl('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrl('lunch'),        TopicNarrow(stream.streamId, 'lunch')),
+        ];
+        checkExpectedNarrows(testCases, streams: [stream]);
+      });
+
+      group('on old topic link, with dot-encoding', () {
+        String mkUrl(operand) {
+          return '#narrow/stream/${stream.name}/topic/$operand';
+        }
+        final testCases = [
+          (mkUrl('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrl('google.2Ecom'), TopicNarrow(stream.streamId, 'google.com')),
+          (mkUrl('google.com'),   null),
+          (mkUrl('topic.20name'), TopicNarrow(stream.streamId, 'topic name')),
+          (mkUrl('stream'),       TopicNarrow(stream.streamId, 'stream')),
+          (mkUrl('topic'),        TopicNarrow(stream.streamId, 'topic')),
+        ];
+        checkExpectedNarrows(testCases, streams: [stream]);
+      });
+
+      group('on old topic link, without realm info', () {
+        final testCases = [
+          ('/#narrow/stream/${stream.name}/topic/topic', TopicNarrow(stream.streamId, 'topic')),
+          ('#narrow/stream/${stream.name}/topic/topic',  TopicNarrow(stream.streamId, 'topic')),
```

Comment:
Similar to the stream case above, this description doesn't really fit, and this pair of cases can just be dropped.